src/parquet/bloom.h

```diff
@@ -0,0 +1,178 @@
+  /// @param len Length of bitset.
+  Bloom(const uint8_t* bitset, uint32_t len);
+
+  Bloom(const Bloom& orig) = delete;
+  virtual ~Bloom();
+
+  // Calculate optimal size according to the number of distinct values and false
+  // positive probability.
+  // @param ndv: The number of distinct values.
+  // @param fpp: The false positive probability.
+  // @return optimal number of bits of given n and p.
+  static uint32_t optimalNumOfBits(uint32_t ndv, double fpp);
+
+  // Determine whether an element exist in set or not.
+  // @param hash the element to contain.
+  // @return false if value is definitely not in set, and true means PROBABLY in set.
+  bool find(uint64_t hash);
+
+  // Insert element to set represented by bloom bitset.
+  // @param hash the hash of value to insert into bloom filter..
+  void insert(uint64_t hash);
+
+  // Compute hash for int value by using its plain encoding result.
+  // @param value the value to hash.
+  // @return hash result.
+  uint64_t hash(int value);
+
+  // Compute hash for long value by using its plain encoding result.
+  // @param value the value to hash.
+  // @return hash result.
+  uint64_t hash(int64_t value);
+
+  // Compute hash for float value by using its plain encoding result.
+  // @param value the value to hash.
+  // @return hash result.
+  uint64_t hash(float value);
+
+  // Compute hash for double value by using its plain encoding result.
+  // @param value the value to hash.
+  // @return hash result.
+  uint64_t hash(double value);
+
+  // Compute hash for Int96 value by using its plain encoding result.
+  // @param value the value to hash.
+  // @return hash result.
+  uint64_t hash(const Int96 *value);
+
+  // Compute hash for ByteArray value by using its plain encoding result.
+  // @param value the value to hash.
+  // @return hash result.
+  uint64_t hash(const ByteArray *value);
+
+  // Compute hash for Fixed Length Byte Array value by using its plain encoding result.
+  // @param value the value to hash.
+  // @return hash result.
+  uint64_t hash(const FLBA *value, uint32_t len);
+
+  // Write bloom filter to output stream. A bloom filter structure should include
+  // bitset length, hash strategy, algorithm, and bitset.
+  // @param sink output stream to write
+  void writeTo(const std::shared_ptr<OutputStream>& sink);
+
+ private:
+  // Create a new bitset for bloom filter, at least 256 bits will be create.
+  // @param numBytes number of bytes for bitset
+  void initBitset(uint32_t num_bytes);
+  void setMask(uint32_t key, uint32_t mask[8]);
+  void addElement(uint64_t hash);
+  bool contains(uint64_t hash);
+
+  // The number of bytes of bloom filter bitset.
+  uint32_t num_bytes;
+
+  // Hash strategy used in this bloom filter.
+  HashStrategy hash_strategy;
+
+  // Algorithm applied of this bloom filter.
+  Algorithm algorithm;
+
+  // The underlying byte array for bloom filter bitset.
+  std::unique_ptr<uint8_t[]> bitset;
```

Comment:
This is almost always going to be used as a `uint32_t[]`, yes? Can we use that instead of `uint8_t[]`?

Reply:
done.